dali/pipeline/operators/crop/bbox_crop.h

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+#define DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operators/fused/crop_cast_permute.h"
+
+namespace dali {
+
+class BBoxCrop : public Operator<CPUBackend> {
+  static const unsigned int kAttempts = 100;
+  static const unsigned int kBboxSize = 4;
+
+ protected:
+  struct Bounds {
+    explicit Bounds(std::vector<float> &&bounds)
+        : min_(bounds[0]), max_(bounds[1]) {
+      DALI_ENFORCE(min_ >= 0, "Min should be at least 0.0. Received: " +
+                                  std::to_string(min_));
+      DALI_ENFORCE(bounds.size() == 2, "Bounds should be provided as 2 values");
+      DALI_ENFORCE(min_ <= max_, "Bounds should be provided as: [min, max]");
+    }
+
+    bool Contains(float k) const { return k >= min_ && k <= max_; }
+
+    const float min_, max_;
+  };
+
+  struct Rectangle {
+    explicit Rectangle(float left, float top, float right, float bottom)
+        : left_(left),
+          top_(top),
+          right_(right),
+          bottom_(bottom),
+          area_(right * bottom) {}
+
+    bool Contains(float x, float y) const {
+      return x >= left_ && (x - left_) <= right_ && y >= top_ &&
+             (y - top_) <= bottom_;
+    }
+
+    Rectangle ClampTo(const Rectangle &other) const {
```

Comment:
You have to check if the rectangles are overlapping or if you got valid intersection,
that is if left <= right and top <= bottom.
Otherwise you could get non-existing rectangle with positive area.

Reply:
In the logic of the code, pre-requisite for clamping is to have valid IoU so is guaranteed always to be overlap. Clamping will work for any type of overlap, including complete overlap, partial or when the box is fully engulfed by the patch.

Reply:
Added check at line 73 to ensure overlap before clamp